### PR TITLE
Ported all re2j tests

### PR DESCRIPTION
**Pull request by techaddict**
Contains all tests plus bug fixes

### CODE REVIEW

File: src/main/scala/re2s/Inst.scala [hunk 1]
```diff
@@ -143,6 +143,7 @@ object Inst {
     while (i < runes.length) {
       val rune = runes(i)
       Utils.escapeRune(out, rune)
+      i += 1
```
Comment (MasseGuillaume) [inline src/main/scala/re2s/Inst.scala:146]: haha outch.

File: src/test/scala/re2s/ExecTest.scala [hunk 2]
```diff
@@ -0,0 +1,686 @@
...
+  def testRE2(_file: String) = {
+    var file = _file
+    var in   = classOf[ExecTest].getResourceAsStream("/" + file)
+    // TODO(adonovan): call in.close() on all paths.
+    if (file.endsWith(".gz")) {
+      in = new GZIPInputStream(in)
+      file = file.substring(0, file.length - ".gz".length) // for errors
+    }
+    var lineno  = 0
+    val r       = new UNIXBufferedReader(new InputStreamReader(in, "UTF-8"))
```
Comment (MasseGuillaume) [inline src/test/scala/re2s/ExecTest.scala:83]: `scala.io.Source.fromResource(resource: String, classLoader: ClassLoader)`

Comment (densh) [reply src/test/scala/re2s/ExecTest.scala:83]: I'd recommend not to use `scala.io`. It's an old API that's likely to be on its way out (nobody is working on it, and new alternatives like betterfiles don't depend on it).

File: src/test/scala/re2s/ExecTest.scala [hunk 3]
```diff
@@ -0,0 +1,686 @@
...
+            strings.add(q)
+            continue = true
+          }
+          if (!continue) {
+            // Is a regexp.
+            re = null
+            refull = null
+            try re = RE2.compile(q)
+            catch {
+              case e: Throwable =>
```
Comment (MasseGuillaume) [inline src/test/scala/re2s/ExecTest.scala]: `scala.util.control.NonFatal(e)`

File: src/test/scala/re2s/ExecTest.scala [hunk 4]
```diff
@@ -0,0 +1,686 @@
...
+                  nfail += 1
+                  if (nfail >= 100) fail("stopping after " + nfail + " errors")
+                  continue = true
+                }
+            }
+            if (!continue) {
+              val full = "\\A(?:" + q + ")\\z"
+              try refull = RE2.compile(full)
+              catch {
+                case e: Throwable =>
```
Comment (MasseGuillaume) [inline src/test/scala/re2s/ExecTest.scala]: `scala.util.control.NonFatal(e)`

File: src/test/scala/re2s/ExecTest.scala [hunk 5]
```diff
@@ -0,0 +1,686 @@
...
+  }
+
+  test("FowlerRepetition") {
+    testFowler("repetition.dat")
+  }
+
+  private val NOTAB = RE2.compilePOSIX("[^\t]+")
+
+  def testFowler(file: String): Unit = {
+    val in           = classOf[ExecTest].getResourceAsStream("/" + file)
```
Comment (MasseGuillaume) [inline src/test/scala/re2s/ExecTest.scala]: scala.io.Source.fromResource(resource: String, classLoader: ClassLoader)

File: src/main/scala/re2s/Inst.scala (hunk 1 above)
Comment (densh) [reply src/main/scala/re2s/Inst.scala:146]: 😅

File: src/test/scala/re2s/ExecTest.scala (hunk 2 above)
Comment (MasseGuillaume) [reply src/test/scala/re2s/ExecTest.scala:83]: Make sense!

File: src/test/scala/re2s/UnicodeTest.scala [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+          fail("MIN_FOLD=%#U should be %#U".format(Unicode.MIN_FOLD, i))
+        last = i
+      }
+      i += 1
+    }
+    if (Unicode.MAX_FOLD != last)
+      fail("MAX_FOLD=%#U should be %#U".format(Unicode.MAX_FOLD, last))
+  }
+
+  // TODO(adonovan): tests for:
```
Comment (densh) [inline src/test/scala/re2s/UnicodeTest.scala]: Was this a TODO in original code? We should drop those. TODOs are a giant code smell. We should use tickets instead.

File: src/test/scala/re2s/GoTestUtils.scala [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+    try new String(b, "UTF-8")
+    catch {
+      case e: UnsupportedEncodingException =>
+        throw new IllegalStateException("can't happen")
+    }
+
+  // Convert |idx16|, which are Java (UTF-16) string indices, into the
+  // corresponding indices in the UTF-8 encoding of |text|.
+  //
+  // TODO(adonovan): eliminate duplication w.r.t. ExecTest.
```
Comment (densh) [inline src/test/scala/re2s/GoTestUtils.scala]: Another TODO.

File: src/test/scala/re2s/ExecTest.scala [hunk 8]
```diff
@@ -0,0 +1,686 @@
...
+                nerr += 1
+                continue = true
+              }
+              if (!continue) {
+                // Expand C escapes (a.k.a. Go escapes).
+                if (flag.indexOf('$') >= 0) {
+                  var f = "\"" + field.get(1) + "\""
+                  try field.set(1, Strconv.unquote(f))
+                  catch {
+                    case e: Exception =>
```
Comment (MasseGuillaume) [inline src/test/scala/re2s/ExecTest.scala]: ` scala.util.control.NonFatal(e)`

File: src/test/scala/re2s/ExecTest.scala [hunk 9]
```diff
@@ -0,0 +1,686 @@
...
+                      System.err.println(
+                        "%s:%d: cannot unquote %s\n".format(file, lineno, f))
+                      nerr += 1
+                  }
+                  f = "\"" + field.get(2) + "\""
+                  try field.set(2, Strconv.unquote(f))
+                  catch {
+                    case e: Exception =>
```
Comment (MasseGuillaume) [inline src/test/scala/re2s/ExecTest.scala]: ` scala.util.control.NonFatal(e)`

**Review by densh — CHANGES_REQUESTED** [review]

File: src/test/scala/re2s/ExecTest.scala [hunk 10]
```diff
@@ -0,0 +1,686 @@
...
+                  case e: Exception =>
+                    System.err.println(
+                      "%s:%d: cannot parse result %s\n"
+                        .format(file, lineno, field.get(3)))
+                    nerr += 1
+                    continue = true
+                }
+                //   Field 5: optional comment appended to the report.
+                // Run test once for each specified capital letter mode that we support.
+                if (!continue) {
```
Comment (MasseGuillaume) [inline src/test/scala/re2s/ExecTest.scala]: for complex break code: https://www.scala-lang.org/api/current/scala/util/control/Breaks.html

File: src/test/scala/re2s/ExecTest.scala [hunk 11]
```diff
@@ -0,0 +1,686 @@
...
+                                  "%s:%d: %s.findSubmatchIndex(%s) = %s, want %s\n"
+                                    .format(file,
+                                            lineno,
+                                            pattern,
+                                            text,
+                                            have,
+                                            pos))
+                                nerr += 1
+                                continue = true
+                              }
```
Comment (MasseGuillaume) [inline src/test/scala/re2s/ExecTest.scala]: omg

File: src/test/scala/re2s/ExecTest.scala [hunk 12]
```diff
@@ -0,0 +1,686 @@
...
+                      }
+                    }
+                  }
+                }
+              }
+            }
+          }
+        }
+      }
+    }
```
Comment (MasseGuillaume) [inline src/test/scala/re2s/ExecTest.scala]: omg

**Pull request merged**